pkg/resources/data.go

```diff
@@ -388,6 +605,32 @@ func GetKubernetesCloudProviderName(cluster *kubermaticv1.Cluster) string {
 	}
 }
 
+func ExternalCloudProviderEnabled(cluster *kubermaticv1.Cluster) bool {
+	// If we are migrating from in-tree cloud provider to CSI driver, we
+	// should not disable the in-tree cloud provider until all kubelets are
+	// migrated, otherwise we won't be able to use the volume API.
+	return cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
+		(kubermaticv1helper.ClusterConditionHasStatus(cluster, kubermaticv1.ClusterConditionCSIKubeletMigrationCompleted, corev1.ConditionTrue) ||
+			!metav1.HasAnnotation(cluster.ObjectMeta, kubermaticv1.CSIMigrationNeededAnnotation))
+}
+
+func GetCSIMigrationFeatureGates(cluster *kubermaticv1.Cluster) []string {
+	var featureFlags []string
+	if metav1.HasAnnotation(cluster.ObjectMeta, kubermaticv1.CSIMigrationNeededAnnotation) {
+		// The following feature gates are always enabled when the
+		// 'externalCloudProvider' feature is activated.
+		if cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] {
+			featureFlags = append(featureFlags, "CSIMigration=true", "CSIMigrationOpenStack=true", "ExpandCSIVolumes=true")
+		}
+		// The CSIMigrationNeededAnnotation is removed when all kubelets have
+		// been migrated.
+		if kubermaticv1helper.ClusterConditionHasStatus(cluster, kubermaticv1.ClusterConditionCSIKubeletMigrationCompleted, corev1.ConditionTrue) {
+			featureFlags = append(featureFlags, "CSIMigrationOpenStackComplete=true")
+		}
```

Comment:
FYI: The `CSIMigration*Complete` feature gates are going to be removed in Kubernetes 1.21.

If someone would upgrade a migrated cluster from 1.20 to 1.21, it could break because the feature gate wouldn't exist. Maybe we should just not enable this feature gate for now. It's not that critical anyways.

Reply:
Great! As it does not seem to be mandatory we can remove it, and avoid rolling out some components for nothing.

Reply:
Alternatively, considering that we have the cluster object, we can check the Kubernetes version and:
* If the version is < 1.21, use the `CSIMigrationOpenStackComplete` feature gate,
* Otherwise (if the version is >= 1.21), use the `InTreePluginOpenStackUnregister` feature gate

Reference: https://github.com/kubernetes/kubernetes/pull/98243